src/build.rs

```diff
@@ -674,9 +680,82 @@ extern "C" fn notify_cb(
     .unwrap_or(2)
 }
 
+impl Default for TreeUpdateBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TreeUpdateBuilder {
+    /// Create a new empty series of updates.
+    pub fn new() -> Self {
+        Self {
+            updates: Vec::new(),
+            paths: Vec::new(),
+        }
+    }
+
+    /// Add an update removing the specified `path` from a tree.
+    pub fn remove<T: IntoCString>(&mut self, path: T) -> &mut Self {
+        let path = util::cstring_to_repo_path(path).unwrap();
+        let path_ptr = path.as_ptr();
+        self.paths.push(path);
+        self.updates.push(raw::git_tree_update {
+            action: raw::GIT_TREE_UPDATE_REMOVE,
+            id: raw::git_oid {
+                id: [0; raw::GIT_OID_RAWSZ],
+            },
+            filemode: raw::GIT_FILEMODE_UNREADABLE,
+            path: path_ptr,
+        });
+        self
+    }
+
+    /// Add an update setting the specified `path` to a specific Oid, whether it currently exists
+    /// or not.
+    ///
+    /// Note that libgit2 does not support an upsert of a previously removed path, or an upsert
+    /// that changes the type of an object (such as from tree to blob or vice versa).
+    ///
+    /// The mode given must be one of 0o040000, 0o100644, 0o100755, 0o120000 or
+    /// 0o160000 currently.
+    pub fn upsert<T: IntoCString>(&mut self, path: T, id: Oid, filemode: i32) -> &mut Self {
```

Comment:
The use of i32 here seems a little strange to me.  It seems consistent with the rest of this crate which uses i32 for filemode, but that seems like a strange choice considering *most* unix platforms use u32.  

Since the value is constrained to the git_filemode_t enum, wouldn't it be better to pass in a `FileMode`?  It seems like that would be safer and a clearer.

Reply:
I was going for consistency, but I can very easily change this to `FileMode`.